src/agent/plugins/build/prepare.ts

```diff
@@ -49,69 +50,59 @@ export function beforeJob(executionContext: cm.IExecutionContext, callback) {
-
     //
-    // Get the repo path under the working directory
+    // Get SCM plugin
     //
-    var hashInput = collId + ':' + defId;
-    if (job.environment.endpoints) {
-        job.environment.endpoints.forEach(function (endpoint) {
-            hashInput = hashInput + ':' + endpoint.url;
-        });
-    }
-    // TODO: build dir should be defined in the build plugin - not in core agent
-    var hashProvider = crypto.createHash("sha256");
-    hashProvider.update(hashInput, 'utf8');
-    var hash = hashProvider.digest('hex');
-    var workingFolder = variables[cm.agentVars.workingDirectory];
-    var buildDirectory = path.join(workingFolder, sys, hash);
-
-    executionContext.info('using build directory: ' + buildDirectory);
-
-    job.environment.variables['agent.buildDirectory'] = buildDirectory;
-    shell.mkdir('-p', buildDirectory);
-    shell.cd(buildDirectory);
-
-    var repoPath = path.resolve('repo');
-    job.environment.variables['build.sourceDirectory'] = repoPath;
-    job.environment.variables['build.stagingdirectory'] = path.resolve("staging");
-
-    // TODO: remove compat variable
-    job.environment.variables['sys.sourcesFolder'] = repoPath;
-
     var scmm;
     var providerType = endpoint.type.toLowerCase();
     executionContext.info('using source provider: ' + providerType);
 
-    try {
+    try {        
         var provPath = path.join(executionContext.scmPath, providerType);
         executionContext.info('loading: ' + provPath);
-        scmm = require(provPath);    
+        scmm = require(provPath);
     }
     catch(err) {
         callback(new Error('Source Provider not found: ' + providerType));
         return;        
     }
     
-    var scmProvider: cm.IScmProvider = scmm.getProvider(executionContext, repoPath);
-    scmProvider.hash = hash;
-    scmProvider.initialize(endpoint);
+    if (!scmm.getProvider) {
+        throw new Error('SCM Provider does not implement getProvider: ' + providerType);
+    }
+    
+    var scmProvider: cm.IScmProvider = scmm.getProvider(executionContext, endpoint);
+    scmProvider.initialize();
     scmProvider.debugOutput = executionContext.debugOutput;
-
-    return Q(null)
-    .then(() => {
+    var hashKey: string = scmProvider.hashKey;
+    
+    //
+    // Get source mappings and set variables
+    //
+    var workingFolder = variables[cm.vars.agentWorkingDirectory];
+    var repoPath: string;
+    
+    var sm: smm.SourceMappings = new smm.SourceMappings(workingFolder, executionContext.hostContext);
+    sm.getSourceMapping(hashKey, job, endpoint)
+    .then((srcMap: smm.ISourceMapping) => {
+        repoPath = scmProvider.targetPath = path.join(workingFolder, srcMap.build_sourcesdirectory);
+
+        //
+        // Variables
+        //        
+        variables[cm.vars.buildSourcesDirectory] = repoPath;
+        variables[cm.vars.buildArtifactsStagingDirectory] = path.join(workingFolder, srcMap.build_artifactstagingdirectory);
```

Comment:
I thought we were going with ArtifactStagingDirectory with no s as that is the current implementation for the windows agent.  I already have a change staged for the docs.